src/debug_sdfs_ed64.c

```diff
@@ -0,0 +1,590 @@
+		// UVWX YZ?!     NVPX RZT!
+		t = ((x ^ (x >> 7)) & 0x00AA00AA00AA00AA);
+		x = (x ^ t ^ (t << 7));
+
+		// Transpose 2x2 blocks inside their 4x4 blocks in the 8x8 matrix
+		// aick emgo     aiqy emu2
+		// bjdl fnhp     bjrz fnv3
+		// qys0 u2w4     cks0 gow4
+		// rzt1 v3x5  \  dlt1 hpx5
+		// 6E8G AICK  /  6EMU AIQY
+		// 7F9H BJDL     7FNV BJRZ
+		// MUOW QYS?     8GOW CKS?
+		// NVPX RZT!     9HPX DLT!
+		t = ((x ^ (x >> 14)) & 0x0000CCCC0000CCCC);
+		x = (x ^ t ^ (t << 14));
+
+		// collect successive 4bits to be interleaved with their pair
+		// t <- 0000 0000 0000 0000 0000 0000 0000 0000 aiqy 6EMU bjrz 7FNV cks0 8GOW dlt1 9HPX
+		// x <- 0000 0000 0000 0000 0000 0000 0000 0000 emu2 AIQY fnv3 BJRZ gow4 CKS? hpx5 DLT!
+		t = ((x & 0xF0F0F0F000000000) >> 32) | ((x & 0x00000000F0F0F0F0) >> 4);
+		x = (((x & 0x0F0F0F0F00000000) >> 28) | (x & 0x000000000F0F0F0F));
+
+		// interleave 4 bits to form the real bytes
+		x = everdrive_sd_interleave_bits(t, x);
+
+		// At this point x is properly interleaved.
+		//      0       1       2       3       4       5       6      7
+		// x <- [63..56][55..48][47..40][39..32][31..24][23..16][15..8][7..0]
+		//      |----------dat3||----------dat2||----------dat1||-------dat0|
+
+		// For every dat line
+		for (int i = 3; i >= 0; i--) {
+			tx = dat_crc[i] >> 8 ^ (x >> (i * 16 + 8));
+			tx ^= tx >> 4;
+			dat_crc[i] = (dat_crc[i] << 8) ^ (uint16_t)(tx << 12) ^(uint16_t)(tx << 5) ^ tx;
+
+			tx = dat_crc[i] >> 8 ^ (x >> (i * 16));
+			tx ^= tx >> 4;
+			dat_crc[i] = (dat_crc[i] << 8) ^ (uint16_t)(tx << 12) ^ (uint16_t)(tx << 5) ^ tx;
+		}
+
+		data_p += 8;
+	}
+
+	// The hardware interface will write any given data to the lanes in packed
+	// format so we need to interleave the crc to take a bit from each CRC per
+	// line
+	t = (uint64_t)dat_crc[3] << 32 | dat_crc[2];
+	y = (uint64_t)dat_crc[1] << 32 | dat_crc[0];
+
+	t = everdrive_sd_interleave_bits(t, y);;
+
+	y = t & 0x00000000FFFFFFFF;
+	t = t >> 32;
+	
+	t = everdrive_sd_interleave_bits(t, y);;
+
+	crc_out[0] = t >> 48;
+	crc_out[1] = t >> 32;
+	crc_out[2] = t >> 16;
+	crc_out[3] = t;
+}
+
+static uint8_t everdrive_sd_stop_transmission() {
+	uint8_t result;
+	uint16_t timeout = -1;
+
+	if ((everdrive_sd_active_mode & ED64_SD_MODE_ACCESS) == ED64_SD_MODE_NONE) return 0;
+
+	result = everdrive_sd_execute_command(ED64_SD_CMD12, 0);
+	if (result & ED64_SD_TIMEOUT) return result;
+
+	// Wait until the buffer is cleaned
+	set_everdrive_sd_bitlen(8);
+	while (everdrive_sd_read_data() != 0xff) {
+		if (!timeout--) {
+			debugf("Buffer timed out\n");
+			return 0xFF;
+		}
+	}
+
+	return 0;
+}
+
+static LBA_t everdrive_sd_address = 0;
+static uint8_t everdrive_sd_change_mode(uint32_t mode, LBA_t addr) {
+	uint8_t result;
+	if (
+		(everdrive_sd_active_mode & ED64_SD_MODE_ACCESS) == mode &&
+		everdrive_sd_address == addr
+	) {
+		return 0;
+	}
+
+	result = everdrive_sd_stop_transmission();
+	if (result & ED64_SD_TIMEOUT) return result;
```

Comment:
Why calling stop transmission here? What does that function do?

Reply:
Once a block read/write command is issued, the SD card continuously sends data as the clock is toggled. To exit from this mode, we dispatch a `CMD12`. I will try to make this clearer.